File: Assignments/Week4/matlab/SymMatVec1.m
function [ y_out ] = SymMatVec1( A, x, y )
% Compute y := A x + y, assuming A is symmetric and stored in lower
% triangular part of array A.

% Extract the row and column size of A
[ m, n ] = size( A );

% (Strictly speaking you should check that m = n, x is a vector size n and y is a
% vector of size n...)

% Copy y into y_out
y_out = y;

% Compute y_out = A * x + y_out
for i = 1:n
    for j=1:i
        y_out( i ) = A( i,j ) * x( j ) + y_out( i );
    end
    for j=i+1:n
        y_out( i ) = A( j,i ) * x( j ) + y_out( i );
    end
end

end

File: Assignments/Week4/C/SymMatVec1.c
#define A( i,j ) *( ap + (j)*lda + i )  // map A( i,j ) to array ap in column-major order
#define x( i ) *( xp + i )              // map x( i ) to array xp
#define y( i ) *( yp + i )              // map y( i ) to array yp


void SymMatVec1( int n,
		 double *ap,
		 int lda,
		 double *xp,
		 double *yp )
{
  int i, j;

  for ( i=0; i<n; i++ ){
    for ( j=0; j<=i; j++ )
      y( i ) += A( i, j ) * x( j );
    for ( j=i+1; j<n; j++ )
      y( i ) += A( j, i ) * x( j );
  }

  return;
}
  
